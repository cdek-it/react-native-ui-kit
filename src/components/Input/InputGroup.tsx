import { memo, useCallback, useImperativeHandle, useMemo, useRef } from 'react'
import { type TextInput, View, StyleSheet, type ViewStyle } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'

import { FloatLabel, type FloatLabelProps } from './FloatLabel'
import { InputGroupAddon, type InputGroupAddonProps } from './InputGroupAddon'
import { InputTextBase } from './InputTextBase/InputTextBase'
import type { InputTextBaseProps } from './InputTextBase/types'

/** @see InputTextBaseProps */
interface InputGroupBaseProps {
  /** Содержимое левого аддона группы */
  left?: InputGroupAddonProps['content']
  /** Содержимое правого аддона группы */
  right?: InputGroupAddonProps['content']
  /** Дополнительная стилизация для контейнера компонента */
  style?: ViewStyle
}

export type InputGroupProps =
  | (InputGroupBaseProps & { input: 'FloatLabel' } & FloatLabelProps)
  | (InputGroupBaseProps & { input: 'InputTextBase' } & InputTextBaseProps)

/**
 * Компонент группы инпута. Позволяет обернуть инпут в группу и добавить кнопки-аддоны слева и справа
 * @link https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-5932&m=dev
 * @see InputTextBase
 * @see InputGroupAddon
 */
export const InputGroup = memo<InputGroupProps>(
  ({
    left,
    right,
    style,
    disabled,
    inputRef: propsInputRef,
    input = 'InputTextBase',
    ...otherProps
  }) => {
    const styles = useStyles()
    const inputRef = useRef<TextInput>(null)

    const focus = useCallback(() => inputRef.current?.focus(), [inputRef])
    useImperativeHandle(propsInputRef, () => inputRef.current, [inputRef])

    const containerStyle = useMemo(() => {
      return StyleSheet.flatten([
        styles.inputContainer,
        !!left && styles.inputContainerForLeftAddon,
        !!right && styles.inputContainerForRightAddon,
      ])
    }, [
      left,
      right,
      styles.inputContainer,
      styles.inputContainerForLeftAddon,
      styles.inputContainerForRightAddon,
    ])

    const renderInput = useCallback(() => {
      if (input === 'FloatLabel') {
        return (
          <FloatLabel
            containerStyle={containerStyle}
            disabled={disabled}
            inputRef={inputRef}
            placeholder={otherProps.placeholder || ''}
            {...otherProps}
          />
        )
      }

      return (
        <InputTextBase
          containerStyle={containerStyle}
          disabled={disabled}
          inputRef={inputRef}
          {...otherProps}
        />
      )
    }, [input, inputRef, containerStyle, disabled, otherProps])

    return (
      <View style={[styles.container, style]}>
        {left ? (
          <InputGroupAddon
            content={left}
            disabled={disabled}
            position='left'
            onPress={focus}
          />
        ) : null}

        <View style={styles.inputWrapper}>{renderInput()}</View>

        {right ? (
          <InputGroupAddon
            content={right}
            disabled={disabled}
            position='right'
            onPress={focus}
          />
        ) : null}
      </View>
    )
  }
)

const useStyles = makeStyles(() => ({
  container: { flexDirection: 'row' },
  inputWrapper: { flexGrow: 1, zIndex: 10 },
  inputContainer: { flexGrow: 1 },
  inputContainerForLeftAddon: {
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
  },
  inputContainerForRightAddon: {
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0,
  },
}))
