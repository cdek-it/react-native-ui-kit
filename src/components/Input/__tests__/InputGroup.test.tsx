import { render } from '@testing-library/react-native'

import { InputGroup, type InputGroupProps } from '../InputGroup'

jest.mock('../InputTextBase/InputTextBase', () => ({
  InputTextBase: 'InputTextBase',
}))

jest.mock('../InputGroupAddon', () => ({ InputGroupAddon: 'InputGroupAddon' }))

describe('InputTextBase component functionality tests', () => {
  const snapshotCases: Array<[string, InputGroupProps]> = [
    ['left', { left: 'left text', input: 'InputTextBase' }],
    ['right', { right: 'right text', input: 'InputTextBase' }],
    [
      'left & right',
      { left: 'left text', right: 'right text', input: 'InputTextBase' },
    ],
    ['no addons, style', { style: { margin: 10 }, input: 'InputTextBase' }],
  ]

  test.each(snapshotCases)('%s', (_, props) => {
    const { toJSON } = render(<InputGroup {...props} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
