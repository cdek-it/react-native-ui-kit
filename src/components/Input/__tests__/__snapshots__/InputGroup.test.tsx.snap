// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextBase component functionality tests left & right 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <InputGroupAddon
    content="left text"
    onPress={[Function]}
    position="left"
  />
  <View
    style={
      {
        "flexGrow": 1,
        "zIndex": 10,
      }
    }
  >
    <InputTextBase
      containerStyle={
        {
          "borderBottomLeftRadius": 0,
          "borderBottomRightRadius": 0,
          "borderTopLeftRadius": 0,
          "borderTopRightRadius": 0,
          "flexGrow": 1,
        }
      }
      inputRef={
        {
          "current": null,
        }
      }
    />
  </View>
  <InputGroupAddon
    content="right text"
    onPress={[Function]}
    position="right"
  />
</View>
`;

exports[`InputTextBase component functionality tests left 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <InputGroupAddon
    content="left text"
    onPress={[Function]}
    position="left"
  />
  <View
    style={
      {
        "flexGrow": 1,
        "zIndex": 10,
      }
    }
  >
    <InputTextBase
      containerStyle={
        {
          "borderBottomLeftRadius": 0,
          "borderTopLeftRadius": 0,
          "flexGrow": 1,
        }
      }
      inputRef={
        {
          "current": null,
        }
      }
    />
  </View>
</View>
`;

exports[`InputTextBase component functionality tests no addons, style 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      {
        "margin": 10,
      },
    ]
  }
>
  <View
    style={
      {
        "flexGrow": 1,
        "zIndex": 10,
      }
    }
  >
    <InputTextBase
      containerStyle={
        {
          "flexGrow": 1,
        }
      }
      inputRef={
        {
          "current": null,
        }
      }
    />
  </View>
</View>
`;

exports[`InputTextBase component functionality tests right 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexGrow": 1,
        "zIndex": 10,
      }
    }
  >
    <InputTextBase
      containerStyle={
        {
          "borderBottomRightRadius": 0,
          "borderTopRightRadius": 0,
          "flexGrow": 1,
        }
      }
      inputRef={
        {
          "current": null,
        }
      }
    />
  </View>
  <InputGroupAddon
    content="right text"
    onPress={[Function]}
    position="right"
  />
</View>
`;
